add services guide to manage a swarm

sfsmithcha opened the pull request:
Adds guide for creating services to `Manage a swarm`. Doesn't cover every option/not intended to be comprehensive, but a good start for most command line options. Would like to get something in along w/ 1.12 release.

Signed-off-by: Charles Smith charles.smith@docker.com

sfsmithcha committed:
add services guide to manage a swarm

Signed-off-by: Charles Smith <charles.smith@docker.com>

sfsmithcha commented:
ping @mgoelzer , @stevvooe , @aaronlehmann

sfsmithcha commented:
CI Errors expected until 1.12 tag is out

aaronlehmann commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,252 @@
+$ docker service create <IMAGE-ID>
+```
+
+The swarm orchestrator schedules one task on an available node. The task invokes  a container based upon the image. For example you could run the following command to create a service of one instance of an nginx web server:
````
Extra space after "invokes".

aaronlehmann commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,252 @@
+a container image:
+
+```bash
+$ docker service create <IMAGE-ID>
````
`<IMAGE-REFERENCE>`

Image ID has a specific meaning; reference is more generic.

aaronlehmann replied:
Or `<IMAGE>` for brevity.

aaronlehmann commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,252 @@
+anixjtol6wdf  my_web  1/1       nginx
+```
+
+You can include a command to run inside containers after the image id:
````
"image reference"

aaronlehmann commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,252 @@
+You can include a command to run inside containers after the image id:
+
+```bash
+$ docker service create <IMAGE-ID> <COMMAND>
````
`<IMAGE-REFERENCE>`

aaronlehmann replied:
Or `<IMAGE>`.

aaronlehmann commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,252 @@
+$ docker service create <IMAGE-ID> <COMMAND>
+```
+
+For example to start an `alpine` task that runs `ping docker.com`:
````
"...to start a service based on the `alpine image` that runs..."

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,252 @@
+
+## Configure update behavior
+
+When you create a service, you can specify a rolling update behavior for how the swarm should apply changes to the service when you run `docker service update`.
```
You can also specify these flags as part of the update, as arguments to `docker service update`.

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,252 @@
+
+The `--update-delay` flag configures the time delay between updates to a service task or sets of tasks. You can describe the time `T` as a combination of the number of seconds `Ts`, minutes `Tm`, or hours `Th`. So `10m30s` indicates a 10 minute 30 second delay.
+
+By default the scheduler updates 1 task at a time. You can pass the ``--update-parallelism` flag to configure the maximum number of service tasks that the scheduler updates simultaneously.
```
Extra backtick before --update-parallelism will mess up the formatting.

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,252 @@
+
+You can create two types of mounts for services in a swarm, `volume` mounts or `bind` mounts. You pass the `--mount` flag when you create a service. If you don't specify a type of mount, it defaults to volume.
+
+Volumes are persistent storage that remain once after a container for a task has been removed. The following samples show volume mount syntax:
```
I see that the paragraph on bind mounts below is a list entry (it starts with `*`). Should this one be too?

aaronlehmann replied:
Extra word: "once"

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,252 @@
+
+## Configure Mounts
+
+You can create two types of mounts for services in a swarm, `volume` mounts or `bind` mounts. You pass the `--mount` flag when you create a service. If you don't specify a type of mount, it defaults to volume.
```
"it defaults to a volume mount"?

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,252 @@
+$ docker service create \
+  --mount type=volume,src=<VOLUME-NAME>,dst=<CONTAINER-PATH>,volume-driver=<DRIVER>,volume-opt=<KEY0>=<VALUE0>,volume-opt=<KEY1>=<VALUE2>
+  --name myservice \
+  <IMAGE-ID>
```
`<IMAGE-REFERENCE>` or `<IMAGE>` in all three examples above.

mgoelzer replied:
At the end of the long line, I think you meant `<KEY1>=<VALUE1>` (not `2`)

stevvooe commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,252 @@
+The swarm orchestrator schedules one task on an available node. The task invokes  a container based upon the image. For example you could run the following command to create a service of one instance of an nginx web server:
+
+```bash
+$ docker service create --name my_web nginx
````
If you do something like this:

```
docker service create -p8080:80 nginx
```

Then wait for the service to start, `curl` will work:

```console
$ curl localhost:8080/                                                                                                                                                                                                                                         ⏎ increase-observation-weight ✭ ✱ ◼
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
```

sfsmithcha replied:
This is intended to demonstrate the most simple command. I've added your example to publish port and added a link here: 

To make the web server acceisble outside the swarm, you need to [publish the port](#publish-ports-externally-to-the-swarm).

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,252 @@
+$ docker service create \
+  --mount type=bind,src=<HOST-PATH>,dst=<CONTAINER-PATH>,readonly \
+  --name myservice \
+  <IMAGE-ID>
```
`<IMAGE-REFERENCE>` or `<IMAGE>` in both examples above.

stevvooe commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,252 @@
+  <IMAGE-ID>
+```
+
+* Bind mounts are file system paths from the host that Docker mounts into the
````
Clarify that this will be the file system path of the node to which the task is scheduled.

stevvooe commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,252 @@
+
+# Same as above
+$ docker service create \
+  --mount type=volume,src=<VOLUME-NAME>,dst=<CONTAINER PATH>
```
Clarify that in this example, we include the type, but the default type is volume.

stevvooe commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,252 @@
+  --name myservice \
+  <IMAGE-ID>
+
+# To create a volume at service creation time
```
It is recommend to use the volume commands, over this method. It should be used if you want the volume to created when the task is dispatched.

The volume is not created when the service is created. It is created before starting up the container associated with the service.

trapier replied:
> It is recommend to use the volume commands, over this method.

Why?

sfsmithcha commented:
Thanks for the great feedback @aaronlehmann , @stevvooe . I think I've addressed your comments now.

@thaJeztah PTAL if you have time.

Would like to get this in for a publish later today.

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,304 @@
+When you are running Docker Engine in swarm mode, you run
+`docker service create` to deploy your application in the swarm. The swarm
+manager accepts the service description as the desired state for your
+application. The built-in swarm orchestrator and scheduler deploy your application to nodes in your swarm to achieve and maintain the desired state.
```
This line is very long and should be wrapped.

aaronlehmann commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,304 @@
+You can include a command to run inside containers after the image:
+
+```bash
+$ docker service create <IMAGD> <COMMAND>
````
IMAGE is misspelled.

aaronlehmann commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,304 @@
+```
+
+Service constraints let you set criteria for a node to meet before the scheduler
+deploys a replicated service to the node. You can apply constraints to the
````
Constraints also work for global services.

sfsmithcha replied:
This seems weird to me. If they are subject to constraints. They are not global.

aaronlehmann replied:
> This seems weird to me. If they are subject to constraints. They are not global. :D But fixing now, that's why we have review.

I remember at one point we thought about calling them "fill" services or "cover" services, as in they cover the set of nodes subject to constraints. I don't remember how we settled on "global". Probably we felt that the alternatives were unclear. As you can imagine, it was quite the bikeshed.

cc @dongluochen

aaronlehmann replied:
Also, every time I say "global services", I think of this: http://www.wsj.com/articles/SB10001424127887323423804579025120455867410

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,304 @@
+
+The scheduler will deploy nginx tasks to a maximum of 3 nodes. However, the
+swarm makes nginx port 80 from the task container accessible at port 8080 on any
+node in the swarm. You can use `curl` on any node to access the web server:
```
"You can direct `curl` at port 8080 of any node"?

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,304 @@
+to the network.
+
+When you create a service and pass the `--network` flag to deploy the service to
+the overlay network in the swarm:
```
"to attach the service to the overlay network"?

aaronlehmann commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,304 @@
+716thylsndqma81j6kkkb5aus
+```
+
+The swarm extends the `my-network` to each node running the service.
````
extra `the`?

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,304 @@
+--update-parallelism` flag to configure the maximum number of service tasks that
+the scheduler updates simultaneously.
+
+By default, when an update to an individual task returns a state of `RUNNING`, the scheduler schedules another task to update until all tasks are updated. If, at any time during an update a task returns `FAILED`, the scheduler pauses the update. You can control the behavior using the `--update-failure-action` flag for docker service create or docker service update.
```
Long line should be wrapped.

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,304 @@
+minute 30 second delay.
+
+By default the scheduler updates 1 task at a time. You can pass the
+--update-parallelism` flag to configure the maximum number of service tasks that
```
Oops, there was a double backtick here before before, but it looks like you removed both, so it's still not right.

sfsmithcha commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,323 @@
+  <IMAGE>
+```
+
+>**Important note:** Bind mounts can be useful but they are also dangerous.  In most cases, we recommend that you architect your application such that mounting paths from the host unnecessary. The main risks include the following:<br><br>
````
@stevvooe , PTAL

aaronlehmann replied:
This line should be wrapped.

aaronlehmann commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,323 @@
+> containers at any time if they become unhealthy or unreachable.<br><br>
+> Host bind mounts are completely non-portable.  When you use  bind mounts,
+> there is no guarantee that your application will run the same way in
+> development as it does in production.
```
It seems a bit strange to have these `<br>` tags here when we don't have them anywhere else in the document, but maybe this this goes beyond my knowledge of markdown.

If we keep them, shouldn't they be `<br />` to be XHTML compliant?

sfsmithcha replied:
Updated to use `<br />` . It is the only way I've found to get the paragraphs to work right inside a note block. 
![note](https://cloud.githubusercontent.com/assets/2415653/17258841/a3d01066-557d-11e6-9bf5-6be6670242df.png)

aaronlehmann commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,326 @@
+  <IMAGE>
+```
+
+>**Important note:** Bind mounts can be useful but they are also dangerous.  In most cases, we recommend that you architect your application such that mounting paths from the host unnecessary. The main risks include the following:<br />
````
This line seems a lot longer than all the others and should probably be wrapped. But if that causes problems with the rendering, don't worry about it.

aaronlehmann commented:
LGTM

sfsmithcha commented:
CI Errors are unrelated to this PR:
```
18:27:40 ERROR: 2016/07/29 18:27:40 content_renderer.go:56: LinkResolver: No page found for "/docker-for-mac/index.md" on page "machine/overview.md".
18:27:40 ERROR: 2016/07/29 18:27:40 content_renderer.go:56: LinkResolver: No page found for "/docker-for-windows/index.md" on page "machine/overview.md".
18:27:40 ERROR: 2016/07/29 18:27:40 content_renderer.go:56: LinkResolver: No page found for "/docker-for-mac/index.md" on page "machine/overview.md".
18:27:40 ERROR: 2016/07/29 18:27:40 content_renderer.go:56: LinkResolver: No page found for "/docker-for-windows/index.md" on page "machine/overview.md".
18:27:40 ERROR: 2016/07/29 18:27:40 content_renderer.go:56: LinkResolver: No page found for "/docker-for-mac/index.md" on page "toolbox/overview.md".
18:27:40 ERROR: 2016/07/29 18:27:40 content_renderer.go:56: LinkResolver: No page found for "/docker-for-windows/index.md" on page "toolbox/overview.md".
```

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+
+For an overview of how services work, refer to [How services work](how-swarm-mode-works/services.md).
+
+This guide assumes you are working with the Docker Engine in running in swarm
```
in running in -> running in

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+mode. You must run all `docker service` commands from a manager node.
+
+If you haven't already, read through [Swarm mode key concepts](key-concepts.md)
+and [How services work](how-swarm-mode-works/services.md).
```
nit: put comma before "and" (we standardize on "oxford comma") https://docs.docker.com/opensource/doc-style/#/oxford-comma

aaronlehmann replied:
meta-nit: The Oxford comma only applies to lists of 3 or more items. I don't believe there should be a comma in a list of two items.

sfsmithcha replied:
no oxford comma in a 2 item list

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+
+## Create a service
+
+To create the simplest type of service in a swarm, you only only need to supply
```
only only -> only

thaJeztah commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,328 @@
+a container image:
+
+```bash
+$ docker service create <IMAGE>
````
hm, thinking; do we explain somewhere that services always run in "detached" mode, so there's no `-d` / `--detach`?

sfsmithcha replied:
not sure that belongs at this spot at the moment. filed https://github.com/docker/docker/issues/25248 to track.

thaJeztah commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,328 @@
+```
+
+The swarm orchestrator schedules one task on an available node. The task invokes
+a container based upon the image. For example you could run the following
````
comma after "For example"

thaJeztah commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,328 @@
+anixjtol6wdf  my_web  1/1       nginx
+```
+
+To make the web server example accessible from outside the swarm, you need to
````
suggestion: remove "example"

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+
+You publish service ports externally to the swarm using the `--publish
+<TARGET-PORT>:<SERVICE-PORT>` flag. When you publish a service port, the swarm
+makes the service accessible at the target port on any node regardless if there
```
any node, or "every node"?

thaJeztah commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,328 @@
+web server:
+
+```bash
+$ curl localhost:8080/
````
nit: I think we should remove the trailing `/`

stevvooe replied:
Clarify that the service needs to be running on the same node where the service is.

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+number of seconds `Ts`, minutes `Tm`, or hours `Th`. So `10m30s` indicates a 10
+minute 30 second delay.
+
+By default the scheduler updates 1 task at a time. You can pass the
```
should there be a comma after "By default"?

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+`--update-parallelism` flag to configure the maximum number of service tasks
+that the scheduler updates simultaneously.
+
+By default, when an update to an individual task returns a state of `RUNNING`,
```
just a nit: two consecutive paragraphs starting with "By default", probably can drop one

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+that the scheduler updates simultaneously.
+
+By default, when an update to an individual task returns a state of `RUNNING`,
+the scheduler schedules another task to update until all tasks are updated. If,
```
hairy; "the scheduler schedules"; perhaps "the scheduler starts" "deploys?"

stevvooe replied:
The "orchestrator creates replacement tasks" or something to that effect.

stevvooe commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
++++
+title = "Deploy services to a swarm"
+description = "Deploy services to a swarm"
+keywords = ["guide, swarm mode, service"]
```
Does this need to be `["guide", "swarm mode", "swarm", "service"]`?

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+the scheduler schedules another task to update until all tasks are updated. If,
+at any time during an update a task returns `FAILED`, the scheduler pauses the
+update. You can control the behavior using the `--update-failure-action` flag
+for docker service create or docker service update.
```
backticks around `docker service create` and `docker service update`

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+update. You can control the behavior using the `--update-failure-action` flag
+for docker service create or docker service update.
+
+In the example service below the scheduler applies updates to a maximum of 2
```
comma after "below"

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+`bind` mounts. You pass the `--mount` flag when you create a service. The
+default is a volume mount if you don't specify a type.
+
+* Volumes are persistent storage that remain after a container for a task has
```
Need to check this: are we always preserving volumes, also if they're un-named?

stevvooe replied:
Yep. That is docker's behavior.

This statement may be a little strong, as they are not always "persistent".

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+
+* Volumes are persistent storage that remain after a container for a task has
+been removed. The preferred method to mount volumes is to leverage an existing
+volume:
```
this is tricky; the `volume create` has to be executed on the node that the service will run on (requires you to log-in / connect to that node, as it cannot be executed from the manager). You can't know which node that'll be, so either you have to create the volume on *every* node, or use a volume driver that is multi-host aware.

This is just the current situation; better volume support is high on our priority list

thaJeztah replied:
We may need to add some information about that though;

- current limitations
- how to schedule a service on the same node that a pre-created volume is on?
- use a "named" volume, and let docker create it automatically when the service is started
- warning that the default volume driver is **not** multi-host aware, so if the container is rescheduled on another node, it will start with a "fresh" / "empty" volume (with the same name)

thaJeztah replied:
oh, I see "creating volume at deployment time" is mentioned below

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+* Bind mounts are file system paths from the host where the scheduler deploys
+the container for the task. Docker mounts the path into the container. The
+file system path must exist before the swarm initializes the container for the
+task.
```
this may not technically be correct (currently); we currently *create* the path if it's missing.

we wanted to change that for `docker run`, so perhaps we should not mention that here, and work on a patch to actually fail if the path does not exist (because that behavior has been the cause of a lot of confusion)

thaJeztah commented on docs/swarm/services.md:
```diff
@@ -0,0 +1,328 @@
+
+>**Important note:** Bind mounts can be useful but they are also dangerous.  In
+most cases, we recommend that you architect your application such that mounting
+paths from the host unnecessary. The main risks include the following:<br />
```
incomplete sentence "... unnecessary"

thaJeztah commented on docs/swarm/services.md:
````diff
@@ -0,0 +1,328 @@
+  --mount type=bind,src=<HOST-PATH>,dst=<CONTAINER-PATH>,readonly \
+  --name myservice \
+  <IMAGE>
+```
````
We should have an overview somewhere of the available options for mount (perhaps in the `docker service create` reference.

(I can try creating a first version for that if you want)

thaJeztah commented:
Thanks @sfsmithcha, left some nits, and questions inline. Some of my questions are probably out of scope for *this* PR, but something we need to address.

thaJeztah commented:
LGTM (perhaps address some of the nits, but no show-stoppers); we can do a follow up once we have input for the bigger changes

sfsmithcha commented:
author merge
